### PR TITLE
Sprint 3/issue 31#internal messaging system between guests and hosts

### PR DESCRIPTION
**Pull request by STR1ND3R**
# Description

Internal messaging for guest and hosts
imagen

### CODE REVIEW

File: app/javascript/controllers/reset_form_controller.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { Controller } from "@hotwired/stimulus"
+
+export default class extends Controller {
+    reset() {
+        this.element.reset()
+    }
+}
```
Comment (MauricioNC) [inline app/javascript/controllers/reset_form_controller.js:7]: Add  new line at the end of your code

File: app/views/chat_rooms/_new_chat_room_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<div class="w-full flex flex-row">
+  <%= form_with(model: @chat_room, remote: true, class: "d-flex", data: { controller: "reset-form", action: "turbo:submit-end->reset-form#reset" } ) do |f| %>
+    <%= f.text_field :name, class: "mx-2", autocomplete: 'off' %>
+    <%= f.submit data: { disable_with: false }, class: "btn bg-violet-500 hover:bg-violet-400 p-2 m-2 rounded-md"%>
+  <% end %>
+</div>
```
Comment (MauricioNC) [inline app/views/chat_rooms/_new_chat_room_form.html.erb:6]: Add  new line at the end of your code

File: app/views/chat_rooms/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        <div id="messages" class="flex flex-col">
+          <%= render @messages, partial: 'messages/message' %>
+        </div>
+      </div>
+
+      <%= render partial: 'messages/new_message_form', class: "w-full space-around"%>
+
+    </div>
+  <% end %>
+</div>
```
Comment (MauricioNC) [inline app/views/chat_rooms/index.html.erb:57]: Add  new line at the end of your code

File: app/views/messages/_message.html.erb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<div class="border rounded-lg w-fit p-2 my-4 <%= @current_user == message.user ? "place-self-end" : "place-self-start" %>">
+  <% unless message.chat_room.is_private %>
+    <h6 class="text-xs "> <%= message.user.email %> </h6>
+  <% end %>
+  <%= message.content %>
+</div>
```
Comment (MauricioNC) [inline app/views/messages/_message.html.erb:6]: Add  new line at the end of your code

File: app/views/messages/_new_message_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<div>
+  <%= form_with(model: [@single_chat_room ,@message], remote: true, class: "flex", data: { controller: "reset-form", action: "turbo:submit-end->reset-form#reset" } ) do |f| %>
+    <%= f.text_field :content, id: 'chat-text', class: "m-2 w-10/12 ", autocomplete: 'off' %>
+    <%= f.submit data: { disable_with: false }, class: "btn bg-violet-500 hover:bg-violet-400 p-2 m-2 rounded-md" %>
+  <% end %>
+</div>
```
Comment (MauricioNC) [inline app/views/messages/_new_message_form.html.erb:6]: Add  new line at the end of your code

**Review by MauricioNC — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -7,7 +7,17 @@ def index
...
   def show
-    
+    @user = User.find(params[:id])
+    @current_user = current_user
+    @chat_rooms = ChatRoom.public_rooms
+    @users = User.all_except(current_user)
+    @chat_room = ChatRoom.new
+    @message = Message.new
+    @chat_room_name = get_name(@user, @current_user)
+    @single_chat_room = ChatRoom.where(name: @chat_room_name).first || ChatRoom.create_private_room([@user, @current_user], @chat_room_name)
```
Comment (MauricioNC) [inline app/controllers/users_controller.rb:17]: If you want to get only one record from the database, you could use `ChatRoom.find_by_name('your_chatroom_name')` that gets the first record that matches with the parameter value.

Use the `where` method is ok but, it is commonly used to get a bunch of records.

File: app/models/message.rb [hunk 7]
```diff
@@ -1,8 +1,14 @@
...
-  belongs_to :receiver, class_name: 'User'
+  belongs_to :chat_room
+  belongs_to :user
+  validates :content, presence: true
+  after_create_commit { broadcast_append_to self.chat_room }
+  before_create :confirm_participant
+
+  def confirm_participant
+    return unless chat_room.is_private
+      is_participant = Participant.where(user_id: user.id, chat_room_id: chat_room.id).first
```
Comment (MauricioNC) [inline app/models/message.rb:10]: Same here, you are getting only one value, so you could rewrite this as:

```ruby
is_participant = Participant.find_by(user_id: user.id, chat_room_id: chat_room.id)
```

**Review by MauricioNC — CHANGES_REQUESTED** [review]

File: app/javascript/controllers/reset_form_controller.js (hunk 1 above)
Comment (STR1ND3R) [reply app/javascript/controllers/reset_form_controller.js]: done

**Review by MauricioNC — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (STR1ND3R) [reply app/controllers/users_controller.rb]: done

File: app/views/chat_rooms/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+      <div class="overflow-y-scroll">
+        <%= turbo_stream_from @single_chat_room %>
+        <div id="messages" class="flex flex-col">
+          <%= render @messages, partial: 'messages/message' %>
+        </div>
+      </div>
+
+      <%= render partial: 'messages/new_message_form', class: "w-full space-around"%>
+
```
[the remark this answers is not in the document]
Comment (STR1ND3R) [reply app/views/chat_rooms/index.html.erb:54]: done

**Review by MauricioNC — APPROVED** [review]

**Review by AnaCarrizales — APPROVED** [review]

**Review by DevLuis123 — APPROVED** [review]
great!!

**Pull request merged**